### PR TITLE
Add ADR-004-QGB_relayer_security

### PR DESCRIPTION
**Pull request by SweeXordious**
ADR for the relayer security issue we have.

### CODE REVIEW

**Review by evan-forbes — COMMENTED** [review]
This is a pretty good start, and after we expand on it a bit we should be good to go as far as beginning the design process.

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:35]: removed

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:72]: true :smile:

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:15]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:56]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:36]: we probably want to update this portion of the ADR now to reflect our current plan to use universal nonce, but I like the wording here that describes the goal of adding synchrony. We have a data commitment window but the contract *is not* aware of it, which is an issue. How do we plan on using universal nonces, why does this fix the data commitment issue, and how does this fit in with the security assumptions the contract is already making?

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:68]: barring something popping up, I think we can update this section now too

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:80]: note to self that we should add the contracts PR here if its up before we merge this, if not tho nbd

**Review by evan-forbes — COMMENTED** [review]
:+1: after we update this with decision and detailed design after the contracts PR is up, this ADR is good to go!

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: perfect!

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: :+1: :100:

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: :+1:

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md:43]: check now please? I added another section for the synchronous approach

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [reply docs/architecture/ADR-004-QGB_relayer_security.md]: check now please?

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (evan-forbes) [inline docs/architecture/ADR-004-QGB_relayer_security.md:96]: Considering the the design changes that we went with is relatively small compared to some of the other options, I'm perfectly fine with not filling in all of these details now. In fact, I think this creates a fantastic opportunity for a future contributor #471 to dig through the code and figure out how it actually works.

**Review by evan-forbes — APPROVED** [review]
nice work! I think we can merge this and focus on the implementation now.

File: docs/architecture/ADR-004-QGB_relayer_security.md
Comment (SweeXordious) [inline docs/architecture/ADR-004-QGB_relayer_security.md:96]:
```suggestion
```

**Pull request merged**
